apps/web/pages/api/auth/signup.ts

```diff
@@ -7,13 +7,94 @@ import { sendEmailVerification } from "@calcom/features/auth/lib/verifyEmail";
 import { IS_CALCOM } from "@calcom/lib/constants";
 import slugify from "@calcom/lib/slugify";
 import { closeComUpsertTeamUser } from "@calcom/lib/sync/SyncServiceManager";
-import { validateUsernameInTeam, validateUsername } from "@calcom/lib/validateUsername";
+import { validateUsernameInToken, validateUsername } from "@calcom/lib/validateUsername";
 import prisma from "@calcom/prisma";
 import { IdentityProvider } from "@calcom/prisma/enums";
 import { MembershipRole } from "@calcom/prisma/enums";
 import { signupSchema } from "@calcom/prisma/zod-utils";
 import { teamMetadataSchema } from "@calcom/prisma/zod-utils";
 
+async function upsertUser(data: {
+  username: string;
+  userEmail: string;
+  orgId?: number | null;
+  hashedPassword: string;
+}) {
+  const { userEmail, username, orgId, hashedPassword } = data;
+  let user = await prisma.user.findFirst({
+    where: { email: userEmail, username, organizationId: orgId },
+    select: { id: true, email: true, name: true, username: true, createdDate: true },
+  });
+  const parentUser = await prisma.user.findFirst({
+    where: { email: userEmail, linkedByUserId: null },
+    select: { id: true, emailVerified: true, linkedUsers: { select: { id: true } } },
+  });
+  if (!user) {
+    user = await prisma.user.create({
+      data: {
+        username,
+        email: userEmail,
+        emailVerified: parentUser?.emailVerified || undefined,
+        password: hashedPassword,
+        identityProvider: IdentityProvider.CAL,
+        ...(orgId
+          ? {
+              organization: {
+                connect: {
+                  id: orgId,
+                },
+              },
+            }
+          : {}),
+        ...(parentUser
+          ? {
+              linkedBy: {
+                connect: {
+                  id: parentUser?.id,
+                },
+              },
+            }
+          : {}),
+      },
+    });
+  } else {
+    await prisma.user.updateMany({
+      where: { email: userEmail, username, organizationId: orgId },
+      data: {
+        username,
+        password: hashedPassword,
+        emailVerified: new Date(Date.now()),
+        identityProvider: IdentityProvider.CAL,
+        ...(orgId
+          ? {
+              organization: {
+                connect: {
+                  id: orgId,
+                },
+              },
+            }
+          : {}),
+        ...(parentUser
+          ? {
+              linkedBy: {
+                connect: {
+                  id: parentUser?.id,
+                },
+              },
+            }
+          : {}),
+      },
+    });
+  }
+  if (parentUser) {
+    await prisma.user.updateMany({
+      where: { id: { in: [parentUser.id].concat(parentUser.linkedUsers.map((user) => user.id)) } },
+      data: { password: hashedPassword },
+    });
+  }
+  return user;
+}
+
```

Comment:
Artificial upsert as we can use upsert for user as requires unique where conditions